background/services/name/resolvers/uns.ts

```
@@ -0,0 +1,84 @@
+ */
+const IsValidUNSDomainName = (s: string): boolean => {
+  const trimmedString = s.trim()
+  let isValidUNSDomain = false
+
+  // Any valid domain name will have a period as part of the string
+  if (trimmedString.lastIndexOf(".") > 0) {
+    const domainExtension = trimmedString.slice(trimmedString.lastIndexOf("."))
+    const supportedUNSDomains = [
+      ".crypto",
+      ".coin",
+      ".wallet",
+      ".bitcoin",
+      ".888",
+      ".nft",
+      ".dao",
+      ".zil",
+      ".x",
+    ]
+
+    if (supportedUNSDomains.includes(domainExtension)) {
+      isValidUNSDomain = true
+    }
+  }
+  return isValidUNSDomain
+}
+
+export default function unsResolver(): NameResolver<"UNS"> {
+  return {
+    type: "UNS",
+    canAttemptNameResolution(): boolean {
+      return true
+    },
+    canAttemptAvatarResolution(): boolean {
+      return false
+    },
+    canAttemptAddressResolution({ name, network }: NameOnNetwork): boolean {
+      return sameNetwork(network, ETHEREUM) && IsValidUNSDomainName(name)
+    },
+
+    async lookUpAddressForName({
+      name,
+      network,
+    }: NameOnNetwork): Promise<AddressOnNetwork | undefined> {
+      // We try to resolve the name using unstoppable domains resolution
+      const resolution = new UnstoppableDomainsResolution()
+      const currency = ETHEREUM.baseAsset.symbol
+      const address = await resolution.addr(name, currency)
+
+      if (address === undefined || address === null) {
+        return undefined
+      }
+
+      return {
+        address,
+        network,
+      }
+    },
+    async lookUpAvatar(
+      addressOrNameOnNetwork: AddressOnNetwork | NameOnNetwork
+    ) {
+      const { network } = addressOrNameOnNetwork
+
+      return {}
+    },
+    async lookUpNameForAddress({
+      address,
+      network,
+    }: AddressOnNetwork): Promise<NameOnNetwork | undefined> {
+      return {
+        name: "",
+        network: undefined,
+      }
```

For these two, if unsupported, the async function should throw an exception (see e.g. [the known contract resolver's `lookUpAvatar` implementation](https://github.com/tallycash/extension/blob/main/background/services/name/resolvers/known-contracts.ts#L22-L24)).

Similarly, if UNS doesn't support reverse resolution, `canAttemptNameResolution` should return `false`.

See https://github.com/tallycash/extension/blob/main/background/services/name/name-resolver.ts for full API docs!